### ISSUE TITLE
can 2 layer model make sense?

### ISSUE THREAD

**Issue by JKHenry520** [ShenDezhou/Open-GPT#2]
I trained a 2 layer model， but when I sample from the model, it's nonsense。how can I get a make sense model。add data? add layers?

**Comment by ShenDezhou** [comment]
Hi JK Henry,

If possible, the following config would make sense.

Given a 1M text, say 'https://raw.githubusercontent.com/karpathy/char-rnn/master/data/tinyshakespeare/input.txt'.

> gradient_accumulation_steps = 1
> batch_size = 64
> block_size = 256 # context of up to 256 previous characters
> 
> # baby GPT model :)
> n_layer = 6
> n_head = 6
> n_embd = 384
> dropout = 0.2
> 
> learning_rate = 1e-3 # with baby networks can afford to go a bit higher
> max_iters = 5000
> lr_decay_iters = 5000 # make equal to max_iters usually
> min_lr = 1e-4 # learning_rate / 10 usually
> beta2 = 0.99 # make a bit bigger because number of tokens per iter is small
> 
> warmup_iters = 100 # not super necessary potentially

Hope to see your training success.

**Comment by ShenDezhou** [comment]
Hi JK Henry, 

I have committed a notebook to train a babyGPT with reasonable output after 5k steps training.

Here is a detailed writing for your question.

# Readability training

In order to demonstrate the training and inference effects of this project.
Based on Shakespeare's play text 1M[shakespeare_input.txt](raw_books%2Fshakespeare_input.txt), a tokenizer with 1K vocabulary is trained,
And enable the 1K vocabulary during training for data processing, model training, and model reasoning. The switch is use_custom_tokenizer.

The complete training and reasoning tutorials are in [gpt_pretraining_for_beginner_en.ipynb](colab%2Fgpt_pretraining_for_beginner_en.ipynb).

After 10 minutes of training on T4, about 5,000 steps, the model has the ability to write dramas, with a small number of generation flaws, and most of the generation is meaningful.

The training and usage codes of the tokenizer are [train_tiktoken.py](tokenizer%2Ftrain_tiktoken.py) and [use_tiktoken.py](tokenizer%2Fuse_tiktoken.py).

Readers can train tokenizers for other texts by themselves.
